TINKERPOP-967 Support nested-repeat() structures

GCHQResearcher1337 opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-967

This allows nested loops to be used in traversals e.g.
```gremlin
g.V().repeat(out("knows")).until(repeat(out("created")).emit(hasId(3)))
```

NB This has been tested with `docker/build.sh -t -i -n` which passes.

spmallette commented:
wow - this is a cool pull request. it may take some time to go through review as it is a non-trivial change. thanks.

spmallette commented on gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/branch/RepeatTest.java:
```diff
@@ -386,5 +456,26 @@ public void g_V_hasXname_markoX_repeatXoutE_inV_simplePathX_untilXhasXname_rippl
+
+        @Override
+        public Traversal<Vertex, Path> get_g_V_repeatXout_repeatXoutX_timesX1XX_timesX1X_limitX1X_path_by_name() {
+            // NB We need to prevent the RepeatUnrollStrategy from applying to properly exercise this test as this traversal can be simplified
```
Nice to know that `RepeatUnrollStrategy` works in this way. @GCHQResearcher1337 could you add nested `repeat()` traversals (as necessary) to this parameterized unit test:

https://github.com/apache/tinkerpop/blob/d3b6a340dbb9c89355c40089a5c083dc0dbd4c5b/gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/RepeatUnrollStrategyTest.java#L74

to validate that `RepeatUnrollStrategy` continues to behave properly?

spmallette replied:
hmm - seems like we shouldn't have extra barriers. @dkuppitz any thoughts on how this should be handled?

dkuppitz replied:
For now, an extra check seems good to me. In the long run, we should get rid of these NoOpBarrierStep injections altogether as it should only be done by `LazyBarrierStrategy` (that goes back to the whole DFS vs BFS discussion).

spmallette commented:
This PR shows a pretty nice level of understanding of the intricacies of the core of how Gremlin works.  We rarely get PRs that touch this area of TinkerPop to any depth. I found it interesting that most of the plumbing for this feature was already there. It mostly just needed a new `Traverser` species in place to make it all come together. 

I made a separate comment on this PR to expand a unit test case, but other than that, it looks pretty good. As a minor nit, @GCHQResearcher1337 if you could please make a second pass through your changes and `final` any variables you see that can be marked as such (as that is our style). 

@dkuppitz could you take a look at the Gremlin tests added here in detail and see if you would recommend any others?

Anyway, that's my initial pass at this. I expect to give it a second look before I provide my VOTE. I would also like to see feedback from @dkuppitz who is on holiday until next week, so I probably won't have any other feedback until then.

spmallette commented:
@GCHQResearcher1337  I should have also added in my last comment:

1. Please add [CHANGELOG ](https://github.com/apache/tinkerpop/blob/master/CHANGELOG.asciidoc) entries as necessary to 3.4.0
2. This feature is sufficiently interesting that it probably deserves some special attention. Do you mind writing up something to introduce this change to users in the [Upgrade Documentation](https://github.com/apache/tinkerpop/blob/master/docs/src/upgrade/release-3.4.x.asciidoc) for 3.4.0. 

If you like, I can handle the upgrade docs for you when we go to merge. Just let me know how you'd like to proceed.

spmallette commented:
@GCHQResearcher1337 any reason you didn't implement the labelling for `repeat()` as described in the JIRA issue:

```text
repeat('a',out('knows').repeat('b',out('parent')))
```

As the issue states, it's usage would probably be rare, but marko had that intention when he created the ticket so I thought I'd ask about it (also marko has been lurking about and mentioned it to me last night :smile: ).

GCHQResearcher1337 commented:
@spamallete I think I steered away from that because I initially misunderstood what marko meant - as you pointed out the plumbing (`incrLoops` taking a `stepLabel`) for nested repeats is already there, but I didn't find any plumbing had been done for this (neither `times()` nor `loop()` can take a `stepLabel`) so I thought that the 'work on `LoopStep`' had been abandoned. 

Re-reading this I think I see what is involved and can implement this - both `TimesModulating` and `LoopStep` as well as the `loops()` method of a traverser need to be able to take a stepLabel. We would need to maintain a separate user-defined stepLabel that could be accessed.

I can see why `loops('a')` could be helpful, but I don't know why you would want to do `times('a',2)`. Could you please provide an example of this?

PS Hi Marko! Thanks for Gremlin!

spmallette commented:
Some more notes from Marko (i haven't had a chance to let them sink in myself, just pasting them here): 

```text
times(a,2) == loops(a,eq(2))
```

the idea being, you might want to check on the loop count of an outer loop within an inner loop and vice versa. Basically, you are trying to solve:

```text
for(i=0;i<10) { for(j=0;j<10) if(i == x) …. } ….
```

the point being, you want to be able to check for `i` and `j` counters not just at the loop check point but anywhere in the respective body.

And note that naming `repeat()`s is not the same as naming the `repeat()`-step. It isn’t a step label……… though, some thinking on that might be worth it. In essence `repeat(‘a’,…) != repeat(…).as(‘a’)`
the ‘a’ in the first is for the loop stack, not for the step label….but again, perhaps there is a reason to unite the two concepts. Though if done so, a new syntax would be introduced which would add an “irregularity” to Gremlin

spmallette commented:
> perhaps there is a reason to unite the two concepts.

can't really think of a reason to "unite the two concepts" - I think that we want to keep it as `repeat(‘a’,…) != repeat(…).as(‘a’)`. @dkuppitz any thoughts on that?

dkuppitz commented:
I agree it should be treated differently. Since `as()` labeling has an impact on the path history, it really shouldn't be used for this case.

GCHQResearcher1337 commented:
I've now added support for named loops - described in https://issues.apache.org/jira/browse/TINKERPOP-967, however I have not introduced a new `times()` step. 

I think a new `times()` step would be of marginal benefit and could be confusing - currently `times()` is a step modulator on `repeat()`. 

I've also changed to explicit loop stack initialisation rather than setting it up on the increment.  Setting up the stack/counter on the increment creates a problem in the nested context - when the inner `loops()` is called it incorrectly returns the top counter from the stack.

spmallette commented:
Nice @GCHQResearcher1337 - VOTE +1

robertdale commented:
VOTE +1

GCHQResearcher1337 commented on gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/traverser/B_LP_NL_O_P_S_SE_SL_Traverser.java:
```diff
@@ -0,0 +1,156 @@
+public class B_LP_NL_O_P_S_SE_SL_Traverser<T> extends B_LP_O_P_S_SE_SL_Traverser<T> {
+
+    protected Stack<LabelledCounter> nestedLoops;
+    protected ReferenceMap loopNames = null;
```
I'm using a map with weak values to simplify the `resetLoops()` call so that I don't need to walk the Map or store the 'loop name' multiple times.

dkuppitz replied:
So, for my own education, does that mean that `ReferenceMap` automagically removes entries from the map, when there's no more reference to the value (e.g. when you pop a labeled counter from the stack)?

GCHQResearcher1337 replied:
Yep :) magic. I think at garbage collection the entry will get removed, or there will be on access checks of the validity of the reference. I've used something similar before (can't quite remember what) but didn't want to introduce an extra dependency, but luckily I found ReferenceMap in org.apache.commons.collections which was already included and seems to do this. 

I wasn't quite sure if these references would be safe to 'attaching' as I don't fully understand this, but I think attaching only moves a traverser to a different part of the graph so the References/counters should still be safe.
https://github.com/apache/tinkerpop/blob/a80eb84169048ed74c5ad27ebc4d12944fd0136a/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/util/Attachable.java#L62

spmallette replied:
Could you explain a bit further why a weak reference is desirable here? You mentioned that it helps simplify `resetLoops()` but I'm unfortunately not making the connection as to why even with the phraseology you followed with about not needing "to walk the Map or store the 'loop name' multiple times."

GCHQResearcher1337 replied:
Sure. So the loop stack is a stack of LabelledCounters which store the stepId and the loop counter. As per the second part of https://issues.apache.org/jira/browse/TINKERPOP-967 we want to be able to 'name' a loop counter. I could store this in the LabelledCounter alongside the stepId and the counter, but to get more constant time lookup by name I use a separate Map from loop name -> LabelledCounter.

When `resetLoops()` is called the counter is removed from the Stack and also needs to be removed from this lookup Map. Because the map is keyed on 'name' not stepId, I don't know which entry to remove without either:
* Checking every entry on the map to see if the LabelledCounter matches the one I'm removing from the stack.
* Storing the 'name' in the stack of loop counter

A Map with 'Weak values' only holds onto entries while the value (the reference to the LabelledCounter) is valid. When it gets released the entry (the 'name' and the Reference to the LabelledCounter) get removed. This avoids needing a second Map/second index/extra to store lookups from  

I may have overcomplicated this. I was trying to avoid looping over the loopStack (which I expect to be small) and to keep storage down (which is hard to do) - especially considering the existing code seems to want to shave as many bytes as possible off the looping storage (using a short under the hood when the interface exposes an int):
https://github.com/apache/tinkerpop/blob/a80eb84169048ed74c5ad27ebc4d12944fd0136a/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/traverser/B_O_S_SE_SL_Traverser.java#L34

dkuppitz replied:
Hmm, to be realistic, I don't expect more than 2 or 3 named repeat steps in a single traversal, so an index for the names might really not be necessary (IMO).

GCHQResearcher1337 commented on gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/traverser/util/LabelledCounter.java:
```diff
@@ -0,0 +1,85 @@
+public class LabelledCounter implements Serializable, Cloneable {
+
+    private final String label;
+    private final MutableShort count = new MutableShort();
```
Whoops that my mistake. I'll change it to a primitive short. 

It's a short to match the loop counters used elsewhere e.g.
https://github.com/GCHQResearcher1337/tinkerpop/blob/a80eb84169048ed74c5ad27ebc4d12944fd0136a/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/traverser/B_O_S_SE_SL_Traverser.java#L34

But I originally used a Pair<> before switching to a class, which prevented me from using primitive types.

GCHQResearcher1337 commented:
@dkuppitz Thanks for your comments. I think I've fixed the issues you spotted.

dkuppitz commented on gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/branch/RepeatStepTest.java:
```diff
@@ -39,7 +39,9 @@
                 __.out().emit().repeat(out()).times(3),
-                __.repeat(out()).until(hasLabel("x"))
+                __.repeat(out()).until(hasLabel("x")),
+                __.repeat("a", __.out()).times(3),
```
Sorry, perhaps I should have been more clear on what kind of traversals to add here.

Can you please also add `__.repeat(out()).times(3)`? Without the changes in `hashCode()` this traversal would have had the same hashCode as `__.repeat("a", out()).times(3)` and thus this test would have failed.

I guess these two will still collide:

```
__.repeat("a", __.repeat("b", out()))
__.repeat("b", __.repeat("a", out()))
```

...but I can't think of a solution that doesn't involve any expensive method calls.

GCHQResearcher1337 replied:
I don't quite understand your first point - `__.repeat(out()).times(3)` is the first entry in the array and I believe this did fail before the `hashCode()` fix.

Good point about the collision. 
Maybe I could use the same method the emit first and until first use? ie:
```
@Override
    public int hashCode() {
        int result = super.hashCode() ^ (this.repeatTraversal.hashCode() << 1);
        result ^= Boolean.hashCode(this.untilFirst);
        result ^= Boolean.hashCode(this.emitFirst) << 1;
        if (this.loopName != null)
            result ^= this.loopName.hashCode();
        if (this.untilTraversal != null)
            result ^= this.untilTraversal.hashCode();
        if (this.emitTraversal != null)
            result ^= this.emitTraversal.hashCode();
        return result;
    }
```

dkuppitz replied:
Ah, alright, I totally missed the first entry, all good then.

We would have to shift or rotate the `loopName`'s hashCode value by the depth of the repeat traversal. Unfortunately, we don't have this information and I don't see how your `hashCode` implementation above would solve the problem. Anyway, it's an unlikely corner case and I don't think we need to care about it too much, so no big deal if we don't have a solution for now.

dkuppitz commented:
VOTE: +1

spmallette commented:
I will take care of merging this now that we have reviews done. Thanks for patiently working through this process @GCHQResearcher1337 - is that how you will forever be known to us by the way? :smile:

GCHQResearcher1337 commented:
@spmallette Yep. Awesome - thanks for all your input + reviews!
